### KNN Targil 4 -- Shua Friedman


### Part 1

In [1]:
import math
import csv


Q1 and Q2

In [2]:
point = [1, 0, 0, '?']
data1 = [1, 1, 1, 'M']
data2 = [1, 2, 0, 'F']

print(f'The vector {data1[:-1]} has tag {data1[-1]}')
print(f'The vector {data2[:-1]} has tag {data2[-1]}')

The vector [1, 1, 1] has tag M
The vector [1, 2, 0] has tag F


Q3

In [3]:
def euclideanDistance(a, b, length):
   distance = 0
   for x in range(length):
         n_1=float(a[x])
         n_2=float(b[x])
         distance += pow(n_1- n_2, 2)
   return math.sqrt(distance)

In [4]:
#calculate distance between data1 and data2
length = len(data1)-1
distance = euclideanDistance(data1, data2, length)
print(f'The distance between {data1[:-1]} and {data2[:-1]} is {distance}')

The distance between [1, 1, 1] and [1, 2, 0] is 1.4142135623730951


Q4a

In [5]:
with open('myFile.csv', 'r') as myCsvfile:
    lines = csv.reader(myCsvfile)
    dataWithHeader = list(lines)
dataset = dataWithHeader[1:]

In [6]:
for i in range(2):
    print(f'The vector {dataset[i][:-1]} has tag {dataset[i][-1]}')

The vector ['0', '1', '2'] has tag F
The vector ['1', '5', '6'] has tag F


Q4b

In [7]:
#distance btween the first two vectors
distance = euclideanDistance(dataset[0][:-1], dataset[1][:-1], 3)
print(f'The distance between the first two vectors is {distance}')

The distance between the first two vectors is 5.744562646538029


Q5

In [8]:
class distClass:
    dist = None #distance of current point from test point
    tag = None #tag of current point

In [9]:
eucDistances = [] 
point = dataset[0]
for i in range(len(dataset)):
    if i == 0:
        continue
    temp=dataset[i]
    label=temp[-1]
    temp = temp[:-1]
    length = len(temp)
    d=euclideanDistance(point,temp,length)
    print("The distances between " + str(point) + " and " + str(temp) + " is " + str(d))
    print(" and the label is " + label)
    obj = distClass() #one record's distance and tag
    obj.dist=d
    obj.tag=label
    eucDistances.append(obj)

The distances between ['0', '1', '2', 'F'] and ['1', '5', '6'] is 5.744562646538029
 and the label is F
The distances between ['0', '1', '2', 'F'] and ['1', '40', '1'] is 39.02563260217571
 and the label is M
The distances between ['0', '1', '2', 'F'] and ['2', '2', '2'] is 2.23606797749979
 and the label is F
The distances between ['0', '1', '2', 'F'] and ['0', '1', '20'] is 18.0
 and the label is M
The distances between ['0', '1', '2', 'F'] and ['1', '5', '6'] is 5.744562646538029
 and the label is M
The distances between ['0', '1', '2', 'F'] and ['1', '4', '1'] is 3.3166247903554
 and the label is M
The distances between ['0', '1', '2', 'F'] and ['2', '22', '2'] is 21.095023109728988
 and the label is F
The distances between ['0', '1', '2', 'F'] and ['2', '23', '2'] is 22.090722034374522
 and the label is M
The distances between ['0', '1', '2', 'F'] and ['2', '24', '2'] is 23.08679276123039
 and the label is M


Q6

In [10]:
eucDistances.sort(key=lambda x: x.dist)


Q7

In [11]:
for i, x in enumerate(eucDistances):
    print(f"vector {dataset[i]} has distance {x.dist} and tag {x.tag}")

vector ['0', '1', '2', 'F'] has distance 2.23606797749979 and tag F
vector ['1', '5', '6', 'F'] has distance 3.3166247903554 and tag M
vector ['1', '40', '1', 'M'] has distance 5.744562646538029 and tag F
vector ['2', '2', '2', 'F'] has distance 5.744562646538029 and tag M
vector ['0', '1', '20', 'M'] has distance 18.0 and tag M
vector ['1', '5', '6', 'M'] has distance 21.095023109728988 and tag F
vector ['1', '4', '1', 'M'] has distance 22.090722034374522 and tag M
vector ['2', '22', '2', 'F'] has distance 23.08679276123039 and tag M
vector ['2', '23', '2', 'M'] has distance 39.02563260217571 and tag M


Q8

In [12]:
#nearest neighbors with k = 1 to poin
print(f"With K=1, the tag for vector 'point' = : {eucDistances[0].tag}")

With K=1, the tag for vector 'point' = : F


Q9

In [13]:
import numpy as np
k = 3
top3 = [eucDistances[i].tag for i in range(k)]
x, i = np.unique(top3, return_counts=True)
top_tag = x[np.argmax(i)]
print(top3)
print(f"With k = {k}, the tag for vector 'point' = : {top_tag}")


['F', 'M', 'F']
With k = 3, the tag for vector 'point' = : F


#### Part 2 

Q1

In [14]:
import pandas as pd
from sklearn.metrics import accuracy_score

In [15]:
#Gather the code from above into a function for easy use
def knn_func(k,dataset,point, distance_func): #here we pass the distance function as a parameter
  eucDistances = [] 
  for i in range(len(dataset)):
    temp=dataset[i]
    label=temp[-1]
    temp = temp[:-1] #remove the tag
    length=len(temp)
    d=distance_func(point, temp, length)
    obj = distClass() #one record's distance and tag
    obj.dist=d
    obj.tag=label
    eucDistances.append(obj)
    
  eucDistances.sort(key=lambda x: x.dist)
  top_k = [eucDistances[i].tag for i in range(k)]
  x, i = np.unique(top_k, return_counts=True)
  top_tag = x[np.argmax(i)]

  return top_tag


In [16]:
with open('myFile.csv', 'r') as myCsvfile2:
    lines = csv.reader(myCsvfile2)
    dataWithHeader = list(lines)

dataWithHeader[1][3] = 'F'
dataWithHeader[2][3] = 'M'

with open('myFile_test.csv', 'w', newline='') as myCSVtest:
    writer = csv.writer(myCSVtest)
    writer.writerows(dataWithHeader)

In [17]:
k = 3
train = pd.read_csv('myFile.csv')
test = pd.read_csv('myFile_test.csv')

x_test = test.iloc[:,:-1] #divide the test data into x and y
y_test = test.iloc[:, -1]
y_pred = [] #list to store the predictions
for row in x_test.iterrows():
    index, data = row
    y_pred.append(knn_func(k,train.values,data, euclideanDistance)) #perform knn on each row of test data with euclidean distance

#get the accuracy
print(f'Accuracy for k = {k} is: ')
print(accuracy_score(y_test, y_pred))

#replace the label in test with the predictions
test.iloc[:, -1] = y_pred
test.to_csv(f'myFile_test.csv', index=False)

Accuracy for k = 3 is: 
0.6


Q2 a-d: Values of K for Euclidean Distance

In [18]:
#cycle through different values of k, and run the knn function
for k in [1,7,19]:
    train = pd.read_csv('mytrain.csv')
    test = pd.read_csv('mytest.csv')
    x_test = test.iloc[:,:-1]
    y_test = test.iloc[:, -1]
    y_pred = []
    for row in x_test.iterrows():
        index, data = row
        y_pred.append(knn_func(k,train.values,data, euclideanDistance))

    #get the accuracy
    print(f'Accuracy for k = {k} is: ')
    print(accuracy_score(y_test, y_pred))

    #replace the label in test with the predictions
    test.iloc[:, -1] = y_pred
    test.to_csv(f'mytest{k}e.csv', index=False)

Accuracy for k = 1 is: 
0.5
Accuracy for k = 7 is: 
0.74
Accuracy for k = 19 is: 
0.68


    The most accurate value of k is = 7

Q2e: Values of K for Manhattan Distance

In [19]:
def manhattan_distance(a, b, length):
    distance = 0
    for i in range(length):
        distance += abs(a[i] - b[i])
    return distance


In [20]:
from sklearn.metrics import accuracy_score
for k in [1,7,19]:
    train = pd.read_csv('mytrain.csv')
    test = pd.read_csv('mytest.csv')
    x_test = test.iloc[:,:-1]
    y_test = test.iloc[:, -1]
    y_pred = []
    for row in x_test.iterrows():
        index, data = row
        y_pred.append(knn_func(k,train.values,data, manhattan_distance))

    #get the accuracy
    print(f'Accuracy for k = {k} is: ')
    print(accuracy_score(y_test, y_pred))

    #replace the label in test with the predictions
    test.iloc[:, -1] = y_pred
    test.to_csv(f'mytest{k}m.csv', index=False)

Accuracy for k = 1 is: 
0.61
Accuracy for k = 7 is: 
0.63
Accuracy for k = 19 is: 
0.7


Q2f: Values of K for Hamming Distance

In [21]:
def hamming_distance(a,b,length):
    distance = 0
    for i in range(length):
        if a[i] != b[i]:
            distance += 1
    return distance


In [22]:
from sklearn.metrics import accuracy_score
for k in [1,7,19]:
    train = pd.read_csv('mytrain.csv')
    test = pd.read_csv('mytest.csv')
    x_test = test.iloc[:,:-1]
    y_test = test.iloc[:, -1]
    y_pred = []
    for row in x_test.iterrows():
        index, data = row
        y_pred.append(knn_func(k,train.values,data, hamming_distance))

    #get the accuracy
    print(f'Accuracy for k = {k} is: ')
    print(accuracy_score(y_test, y_pred))

    #replace the label in test with the predictions
    test.iloc[:, -1] = y_pred
    test.to_csv(f'mytest{k}h.csv', index=False)

Accuracy for k = 1 is: 
0.61
Accuracy for k = 7 is: 
0.55
Accuracy for k = 19 is: 
0.56


Q2g: Best combination

    The most accurate combination of K and distance function is K=7 and Euclidean Distance. The accuracy is 0.74
